JIT: Fold more nullchecks

EgorBo opened the pull request:
Handles redundant checks in Example 1 in https://github.com/dotnet/runtime/issues/111980

Some [nice diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=934277&view=ms.vss-build-web.run-extensions-tab)

EgorBo committed:
nonnull assertions

EgorBo committed:
fix build

EgorBo committed:
clean up

EgorBo committed:
clean up

EgorBo committed:
clean up

EgorBo committed:
clean up

EgorBo commented:
@AndyAyersMS @jakobbotsch @dotnet/jit-contrib PTAL, [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=934499&view=ms.vss-build-web.run-extensions-tab) show no TP regression.

I also tried to extend the algorithm to handle nested PHIs, but it didn't find additionally anything. Ideally, like Jakob suggested, we should use `VisitReachingVN` function, but it's tricky to extract BBs out of it.
After this PR I want to make a small clean up in various `opt...IsNonNull` functions

EgorBo commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -5810,6 +5833,35 @@ ASSERT_VALRET_TP Compiler::optGetVnMappedAssertions(ValueNum vn)
 }
 
+//------------------------------------------------------------------------
+// optGetEdgeAssertions: Given a block and its predecessor, get the assertions
```
I've extracted this function from rangecheck.cpp to re-use in my function

EgorBo commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -5083,31 +5118,19 @@ bool Compiler::optAssertionVNIsNonNull(ValueNum vn, ASSERT_VALARG_TP assertions)
+    if (!BitVecOps::MayBeUninit(assertions))
     {
-        AssertionIndex assertionIndex = GetAssertionIndex(index);
-        if (assertionIndex > optAssertionCount)
```
redundant condition, I guess it wasn't deleted when we moved all "walk asserts" pieces to `BitVecOps::Iter`.

EgorBo committed:
clean up

EgorBo commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -4367,6 +4367,60 @@ GenTree* Compiler::optAssertionProp_RelOp(ASSERT_VALARG_TP assertions, GenTree*
+            return AssertVisit::Abort;
+        }
+    }
+    return AssertVisit::Continue;
```
NOTE: we can walk nested PHIs (with some budget) or look for PHI defs via SSA longer than the just the first def, but that improved diffs only a tiny bit, so I didn't bother

AndyAyersMS commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -5810,6 +5870,35 @@ ASSERT_VALRET_TP Compiler::optGetVnMappedAssertions(ValueNum vn)
+        return blockPred->bbAssertionOut;
+    }
+
+    if ((blockPred->KindIs(BBJ_ALWAYS) && blockPred->TargetIs(block)) ||
```
I thought for `BBJ_ALWAYS` the right set was `bbAssertionOut`?

AndyAyersMS replied:
Actually for all other BBJ_KINDs you can use `bbAssertionOut`. Just BBJ_COND is special.

EgorBo replied:
@AndyAyersMS are you sure? e.g. what about BBJ_SWITCH or some EH related blocks? I copied this function from rangecheck

EgorBo replied:
I've partially addressed your feedback as I was not confident I can just unconditionally return `bbAssertionOut` for something other than BBJ_ALWAYS/BBJ_COND(false edge). 
However, it looks like we don't typically work with anything but these two anyway.

EgorBo replied:
@AndyAyersMS I see, ok changed

AndyAyersMS commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -5810,6 +5870,35 @@ ASSERT_VALRET_TP Compiler::optGetVnMappedAssertions(ValueNum vn)
+            return bbJtrueAssertionOut[blockPred->bbNum];
+        }
+    }
+    return BitVecOps::UninitVal();
```
Seems like this should return an empty BV on failure, though I also think (once the above is restructured to handle all pred block kinds) you should never get here.

EgorBo replied:
returned empty BV

jakobbotsch commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -4367,6 +4367,60 @@ GenTree* Compiler::optAssertionProp_RelOp(ASSERT_VALARG_TP assertions, GenTree*
+
+    if ((node == nullptr) || !node->OperIs(GT_STORE_LCL_VAR) || !node->Data()->OperIs(GT_PHI))
+    {
+        return AssertVisit::Abort;
```
In fact, what if we use `VNVisitReachingVNs` + just look up assertions from the available assertions in the current block, skipping the stuff around collecting assertions based on edges? I suspect it would work fine.

jakobbotsch replied:
Hmm, it would work for some cases, but not for all cases. E.g.
```csharp
C x = F();
if (y != 0)
  x = G();

if (x != null)
  ...

// VNVisitReachingVNs never visits the phi(F, G) VN, but that's the only VN we have an assertion about

```
So during `VNVisitReachingVNs` we would want to check for assertions on the intermediate phi VNs before proceeding with the phi args. Maybe something for another PR.

However, the first think I suggested (obtaining the phi def via the VN) is probably still worth a try here.

EgorBo replied:
Thanks, addressed!

EgorBo committed:
Address feedback

EgorBo committed:
formatting

jakobbotsch commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -4367,6 +4367,50 @@ GenTree* Compiler::optAssertionProp_RelOp(ASSERT_VALARG_TP assertions, GenTree*
+
+    LclSsaVarDsc*        ssaDef = lvaGetDesc(phiDef.LclNum)->GetPerSsaData(phiDef.SsaDef);
+    GenTreeLclVarCommon* node   = ssaDef->GetDefNode();
+    if ((node == nullptr) || !node->OperIs(GT_STORE_LCL_VAR) || !node->Data()->OperIs(GT_PHI))
+    {
+        return AssertVisit::Abort;
+    }
```
```suggestion
    assert(node->IsPhiDefn());
```

EgorBo committed:
Update src/coreclr/jit/assertionprop.cpp

Co-authored-by: Jakob Botsch Nielsen <Jakob.botsch.nielsen@gmail.com>

EgorBo committed:
Address feedback

EgorBo committed:
Merge branch 'main' into improve-nonnull-assertions

EgorBo commented:
@AndyAyersMS @jakobbotsch anything else? I've addressed your feedback. [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=935176&view=ms.vss-build-web.run-extensions-tab) slightly improved (due to better way to look for PhiDefs via VN).

I plan to use `optVisitReachingAssertions` for a few more things with nice diffs (if not wrong) 🙂

AndyAyersMS commented on src/coreclr/jit/assertionprop.cpp:
```diff
@@ -5810,6 +5859,30 @@ ASSERT_VALRET_TP Compiler::optGetVnMappedAssertions(ValueNum vn)
+{
+    if ((blockPred->KindIs(BBJ_COND) && blockPred->TrueTargetIs(block)))
+    {
+        if (bbJtrueAssertionOut != nullptr)
```
Curious if you ever see this being null.

EgorBo replied:
I assumed it could in rangecheck if assertprop didn't run, but not sure such combination is possible. From a quick SPMI run it's not null. I'll convert it into an assert in my next PR to avoid spinning CI for it

AndyAyersMS reviewed (approved):
LGTM, just one thing I'm puzzled about.

EgorBo commented:
/ba-g "Linux-arm64 is blocked by https://github.com/dotnet/runtime/issues/112021, Build-analysis didn't recognize some failures as known build issue "
